openwisp_controller/tests/utils.py

```diff
@@ -49,3 +54,34 @@ def assertNumQueries(self, num, func=None, *args, using=DEFAULT_DB_ALIAS, **kwar
 
         with context:
             func(*args, **kwargs)
+
+
+class DeviceAdminSeleniumTextMixin(SeleniumTestMixin):
+    def tearDown(self):
+        super().tearDown()
+        # Dismiss any unsaved changes alert to prevent it from blocking
+        # navigation in subsequent tests. If left unresolved, this could
+        # cause test failures by preventing the browser from loading new pages.
+        # Ensuring a clean browser state before the next test runs.
+        try:
+            self.web_driver.refresh()
+        except UnexpectedAlertPresentException:
+            self.web_driver.switch_to_alert().accept()
+            self.web_driver.switch_to_alert.accept()
+        else:
+            try:
+                WebDriverWait(self.web_driver, 1).until(EC.alert_is_present())
+            except TimeoutException:
+                pass
+            else:
+                self.web_driver.switch_to_alert().accept()
+        self.web_driver.refresh()
+        self.wait_for_visibility(By.XPATH, '//*[@id="site-name"]')
+
+    def open(self, url, driver=None):
+        super().open(url, driver)
+        driver = driver or self.web_driver
+        self._assert_loading_overlay_hidden()
+
+    def _assert_loading_overlay_hidden(self):
+        self.wait_for_invisibility(By.CSS_SELECTOR, '#loading-overlay')
```

Comment:
Maybe we want to centralize this in openwisp-utils as I added something similar in openwisp-monitoring, but I had to hide the overlay explicitly as doing this was not enough.